pam/integration-tests/cmd/exec-client/client.go

```diff
@@ -101,34 +99,56 @@ func handleArgs(mTx pam.ModuleTransaction, args []string) error {
 }
 
 func handleArg(mTx pam.ModuleTransaction, arg string) error {
-	// We support handing operations such as:
-	//   <Action>|<input>[|<expected-ret>]
+	// We support handing operations with variant values such as:
+	//   {"action": <"$action">, "args": <[$input...]>, "expects": <[$expected...]>}
 	// Where:
-	//  - <Action> is the method to call
-	//  - <input> is a semicolon-separated list of arguments
-	//  - <expected-ret> is an optional semicolon-separated list of expected return values
+	//  - $action is the method to call
+	//  - $input is a list of variant arguments
+	//  - $expected is an optional list of expected variant return values
 	// For example:
-	//   SetItem|2;user SetItem|-1;foo|29
+	//   {"action": <"SetItem">, "args": <[<2>, <"user">]>}
+	// or:
+	//   {"action": <"SetItem">, "args": <[<-1>, <"foo">]>, "ret": <[<29>]>}
 
 	log.Debugf(context.TODO(), "Parsing argument '%v'", arg)
-	splitArgs := strings.SplitN(arg, "|", 3)
-	action := splitArgs[0]
 
-	var inputArgs *string
-	if len(splitArgs) > 1 && splitArgs[1] != "" {
-		inputArgs = &splitArgs[1]
+	var parsedArg map[string]dbus.Variant
+	variant, err := dbus.ParseVariant(arg, dbus.SignatureOf(parsedArg))
+	if err != nil {
+		return fmt.Errorf("can't parse %s as variant: %w, %w", arg, err, pam_test.ErrInvalidArguments)
 	}
 
-	var expectedRet *string
-	if len(splitArgs) == 3 {
-		expectedRet = &splitArgs[2]
+	err = variant.Store(&parsedArg)
+	if err != nil {
+		return fmt.Errorf("can't store %s as variant map: %w, %w", arg, err, pam_test.ErrInvalidArguments)
+	}
```

Comment:
Nitpick: you can use the shortened-if syntax here

Reply:
I always get worried of loosing the happy path now... :joy: